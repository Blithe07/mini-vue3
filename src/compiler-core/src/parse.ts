import { NodeType } from "./ast"
const enum TagType {
    Start,
    End
}

/**
 * 编译AST解构树
 * @param content 
 * @returns 
 */
export function baseParse(content: string) {
    const context = createParserContext(content)
    return createRoot(parseChildren(context, []))
}
/**
 * 解析子节点
 * @param context 
 * @returns 
 */
function parseChildren(context: Context, ancestors: any[]) {
    const nodes: any = []
    while (!isEnd(context, ancestors)) {
        let node
        const s = context.source
        if (s.startsWith("{{")) {
            node = parseInterpolation(context)
        } else if (s[0] === '<') {
            if (/[a-z]/i.test(s[1])) {
                node = parseElment(context, ancestors)
            }
        } else {
            node = parseText(context)
        }
        nodes.push(node)
    }
    return nodes
}
function isEnd(context: Context, ancestors: any[]) {
    const s = context.source
    if (s.startsWith("</")) {
        for (let i = ancestors.length - 1; i >= 0; i--) {
            const tag = ancestors[i].tag;
            if (startsWithEndTagOpen(s, tag)) {
                return true
            }
        }
    }
    // if (parentTag && s.startsWith(`</${parentTag}>`)) {
    //     return true
    // }

    return !s
}
/**
 * 解析文本
 * @param context 
 * @returns 
 */
function parseText(context: Context) {
    let endIndex = context.source.length
    const endTokens = ["{{", "<"]
    for (let i = 0; i < endTokens.length; i++) {
        const index = context.source.indexOf(endTokens[i])
        if (index > -1 && endIndex > index) {
            endIndex = index
        }
    }
    const content = parseTextData(context, endIndex)
    advanceBy(context, content.length)
    return {
        type: NodeType.TEXT,
        content
    }
}
function parseTextData(context: Context, length: number) {
    return context.source.slice(0, length)
}

/**
 * 处理元素
 * @param context 
 * @returns 
 */
function parseElment(context: Context, ancestors: any[]) {
    let element = parseTag(context, TagType.Start)
    ancestors.push(element)
    element!.children = parseChildren(context, ancestors)
    ancestors.pop()
    if (startsWithEndTagOpen(context.source, element!.tag)) {
        parseTag(context, TagType.End)
    } else {
        throw new Error('缺少结束标签' + element!.tag)
    }
    return element
}
/**
 * 公共抽离方法
 * @param context 
 * @param type 
 * @returns 
 */
function startsWithEndTagOpen(source, tag) {
    return source.slice(2, 2 + tag.length) === tag
}
function parseTag(context: Context, type: TagType) {
    // 解析tag
    const match: any = /^<\/?([a-z]*)/i.exec(context.source)
    // 推进
    advanceBy(context, match[0].length + 1)
    const tag = match[1]
    if (type === TagType.End) return
    return {
        type: NodeType.ELEMENT,
        tag,
        children: []
    }
}
/**
 * 处理插值
 * @param context 
 * @returns 
 */
function parseInterpolation(context: Context) {
    // 抽离变化点
    const openDelimiter = "{{"
    const closeDelimiter = "}}"
    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
    advanceBy(context, openDelimiter.length)
    const rawContentLength = closeIndex - openDelimiter.length
    const rawContent = parseTextData(context, rawContentLength)
    const content = rawContent.trim()
    advanceBy(context, rawContentLength + closeDelimiter.length)
    return {
        type: NodeType.INTERPOLATION,
        content: {
            type: NodeType.SIMPLE_EXPRESSION,
            content
        }
    }
}
/**
 * 推进
 * @param context 
 * @param length 
 */
function advanceBy(context: Context, length: number) {
    context.source = context.source.slice(length)
}
function createRoot(children) {
    return {
        children,
        type:NodeType.ROOT
    }
}
/**
 * 创建上下文对象，用于处理多种解析情况
 * @param content 
 */
function createParserContext(content: string) {
    return {
        source: content
    }
}

type Context = {
    source: string
}