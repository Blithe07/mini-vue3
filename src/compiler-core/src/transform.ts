import { NodeType } from "./ast"
import { TO_DISPLAY_STRING } from "./runtimeHelpers"

export function transform(root, options = {}) {
    const context = createTransformContext(root, options)
    // 1.遍历树，深度优先搜索
    // 2.替换textContent    
    traverseNode(root, context)
    // 生成对应属性给codegen生成代码
    createRootCodegen(root)

    root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root: any) {
    const child = root.children[0]
    if (child.type === NodeType.ELEMENT) {
        root.codegenNode = child.codegenNode
    } else {
        root.codegenNode = root.children[0]
    }
}

function traverseNode(node, context) {
    const nodeTransforms = context.nodeTransforms
    const exitFns: any = []
    for (let i = 0; i < nodeTransforms.length; i++) {
        const transform = nodeTransforms[i];
        const onExit = transform(node, context)
        if (onExit) exitFns.push(onExit)
    }
    switch (node.type) {
        case NodeType.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING)
            break
        case NodeType.ROOT:
        case NodeType.ELEMENT:
            // 获取子节点
            traverseChildren(node, context);
            break
        default:
            break
    }
    let i = exitFns.length
    while (i--) {
        exitFns[i]()
    }
}
function traverseChildren(node: any, context: any) {
    const children = node.children;
    if (children) {
        // 深度优先遍历
        for (let i = 0; i < children.length; i++) {
            const node = children[i];
            traverseNode(node, context);
        }
    }
}

/**
 * 创建上下文对象
 * @param root 
 * @param options 
 */
function createTransformContext(root: any, options: any) {
    const context = {
        root,
        nodeTransforms: options.nodeTransforms || [],
        helpers: new Map(),
        helper(key) {
            context.helpers.set(key, 1)
        }
    }
    return context
}    
