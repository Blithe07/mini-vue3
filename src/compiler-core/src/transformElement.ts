import { createVnodeCall, NodeType } from "./ast";
import { CREATE_ELEMENT_VNODE } from "./runtimeHelpers";

export function transformElement(node, context) {
    if (node.type === NodeType.ELEMENT) {
        return () => {
            const { children, tag } = node
            const vnodeTag = `"${tag}"`
            let vnodeProps
            let vnodeChildren = children[0]
            node.codegenNode = createVnodeCall(context, vnodeTag, vnodeProps, vnodeChildren)
        }
    }
}