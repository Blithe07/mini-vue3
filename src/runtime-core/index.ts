export { h } from './h'
export { renderSlots } from './helper/renderSlots'
export { createTextVnode, createElementVnode } from './vnode'
export { provide, inject } from './apiinject'
export { getCurrentInstance, registerRuntimeCompiler } from './component'
export { nextTick } from './scheduler'
export { toDisplayString } from "../shared"
export * from '../reactivity'