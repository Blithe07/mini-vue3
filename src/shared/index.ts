export const extend = Object.assign
export const isObject = (obj) => {
    return obj !== null && typeof obj === 'object'
}
export const isString = (val) => typeof val === 'string'
export const hasChanged = (val, newVal) => {
    return !Object.is(val, newVal)
}
export const hasOwn = (obj, key) => Object.prototype.hasOwnProperty.call(obj, key)

export const camelize = (str: string) => {
    return str.replace(/-(\w)/g, (_, c: string) => {
        return c ? c.toUpperCase() : ''
    })
}

const capitalize = (str: string) => {
    return str.charAt(0).toUpperCase() + str.slice(1)
}
export const toHandlerKey = (str: string) => {
    return 'on' + capitalize(str)
}

export const EMPTY_OBJ = {}

export * from './toDisplayString'