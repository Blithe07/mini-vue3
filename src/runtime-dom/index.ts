import { createRenderer } from "../runtime-core/renderer"

/**
 * 创建DOM元素
 * @param type 元素类型
 * @returns 
 */
function createElement(type) {
    return document.createElement(type)
}
/**
 * 处理属性高
 * @param el DOM元素
 * @param key 属性名
 * @param oldValue 旧属性值
 * @param newvalue 新属性值
 */
function patchProp(el, key, oldValue, newValue) {
    const isOn = (key: string) => /^on[A-Z]/.test(key)
    if (isOn(key)) {
        // 绑定事件
        // onClick => click
        const event = key.slice(2).toLowerCase()
        el.addEventListener(event, newValue)
    } else {
        if (newValue === null || newValue === undefined) {
            // 移除多余属性
            el.removeAttribute(key)
        } else {
            // 绑定属性
            el.setAttribute(key, newValue)
        }
    }
}
/**
 * 将DOM插入到容器中
 * @param el DOM元素
 * @param container 容器
 * @param anchor 锚点
 */
function insert(el, container,anchor) {
    // container.appendChild(el)
    container.insertBefore(el,anchor)
}
/**
 * 移除DOM
 */
function removeElement(child) {
    const parent = child.parentNode
    if (parent) {
        parent.removeChild(child)
    }
}
/**
 * 设置文本
 */
function setElementText(el, text) {
    el.textContent = text
}
// 得到对象，返回createApp函数
const render = createRenderer({ createElement, patchProp, insert, removeElement, setElementText })

export function createApp(...args) {
    // ...args等同于rootComponent
    return render.createApp(...args)
}

export * from '../runtime-core'