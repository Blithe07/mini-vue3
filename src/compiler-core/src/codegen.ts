import { isString } from "../../shared"
import { NodeType } from "./ast"
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"

/**
 * 只负责生成代码
 * @param ast 
 * @returns 
 */
export function generate(ast) {
    const context = createCodegenContext()
    const { push } = context
    // 导入逻辑，可提取
    genFunctionPreamble(ast, context)
    push("\n")
    push("return ")
    const functionName = 'render'
    const args = ["_ctx", "_cache", "$props", "$setup", "$data", "$options"]
    const signature = args.join(', ')
    push(`function ${functionName}(${signature}) {`)
    push(`return `)
    genNode(ast.codegenNode, context)
    push('}')
    // // 抽离函数名，参数，以及返回值
    // let code = ''
    // code += "return "
    // const functionName = 'render'
    // const args = ["_ctx", "_cache", "$props", "$setup", "$data", "$options"]
    // const signature = args.join(', ')
    // /** 划分职责，抽离节点内容，将节点内容获取交给transform处理 */
    // // const node = ast.children[0]
    // // const node = ast.codegenNode
    // code += `function ${functionName}(${signature}) {
    //         ${genNode(ast, code)}
    //       }`
    return {
        code: context.code
    }
}
function genFunctionPreamble(ast: any, context) {
    const { push, helper } = context
    const VueBinging = "Vue"
    // const helpers = ["toDisplayString"]
    const aliasHelpers = (s) => `${helper(s)}: _${helper(s)}`
    if (ast.helpers.length) {
        push(`const { ${ast.helpers.map(aliasHelpers).join(", ")} } = ${VueBinging}`)
    }
}

function genNode(node, context) {
    switch (node.type) {
        case NodeType.TEXT:
            genText(node, context)
            break;
        case NodeType.INTERPOLATION:
            genInterpolRation(node, context)
            break
        case NodeType.SIMPLE_EXPRESSION:
            genExpression(node, context)
            break
        case NodeType.ELEMENT:
            genElement(node, context)
            break
        case NodeType.COMPOUND_EXPRESSION:
            genCompoundExpression(node, context)
            break
        default: break
    }
}

function genCompoundExpression(node, context) {
    const { push } = context
    const { children } = node
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isString(child)) {
            push(child)
        } else {
            genNode(child, context)
        }
    }
}

function genElement(node, context) {
    const { push, helper } = context
    const { tag, children, props } = node
    push(`_${helper(CREATE_ELEMENT_VNODE)}(`)
    // for (let i = 0; i < children.length; i++) {
    //     const child = children[i];
    //     genNode(child, context)
    // }
    genNodeList(genNullable([tag, props, children]), context)
    // genNode(children, context)
    push(`)`)
}

function genNodeList(nodes, context) {
    const { push } = context
    for (let i = 0; i < nodes.length; i++) {
        const node = nodes[i];
        if (isString(node)) {
            push(node)
        } else {
            genNode(node, context)
        }

        if(i<nodes.length-1){
            push(", ")
        }
    }
}

function genNullable(args) {
    return args.map((arg) => arg || 'null')
}

function genText(node: any, context: any) {
    const { push } = context
    push(`"${node.content}"`)
}

function genInterpolRation(node: any, context: any) {
    const { push, helper } = context
    push(`_${helper(TO_DISPLAY_STRING)}(`)
    genNode(node.content, context)
    push(`)`)
}

function genExpression(node: any, context: any) {
    const { push } = context
    push(`${node.content}`)
}

function createCodegenContext() {
    const context = {
        code: '',
        push(source) {
            context.code += source
        },
        helper(key) {
            return helperMapName[key]
        }
    }
    return context
}
