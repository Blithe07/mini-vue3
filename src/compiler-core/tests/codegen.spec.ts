import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformElement } from "../src/transformElement"
import { transformExpression } from "../src/transformExpression"
import { transformText } from "../src/transformText"

describe('codegen', () => {
    it('text', () => {
        const ast = baseParse('hi')
        transform(ast)
        const { code } = generate(ast)
        // 快照
        expect(code).toMatchSnapshot()
    })
    it('interpolation', () => {
        const ast = baseParse('{{message}}')
        transform(ast, {
            nodeTransforms: [transformExpression]
        })
        const { code } = generate(ast)
        expect(code).toMatchSnapshot()
    })
    it('element', () => {
        const ast: any = baseParse('<div>hi,{{message}}</div>')
        transform(ast, {
            nodeTransforms: [transformExpression, transformElement, transformText]
        })
        const { code } = generate(ast)
        expect(code).toMatchSnapshot()
    })

})