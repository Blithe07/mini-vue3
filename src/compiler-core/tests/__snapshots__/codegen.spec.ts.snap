// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { toDisplayString: _toDisplayString, createElementVnode: _createElementVnode } = Vue
return function render(_ctx, _cache, $props, $setup, $data, $options) {return _createElementVnode(\\"div\\", null, \\"hi,\\" + _toDisplayString(_ctx.message))}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache, $props, $setup, $data, $options) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen text 1`] = `
"
return function render(_ctx, _cache, $props, $setup, $data, $options) {return \\"hi\\"}"
`;
